native/thrift_struct_read_module.c

```diff
@@ -0,0 +1,1851 @@
+        *size = readJsonInteger(data);
+    }
+    if (!strcmp(proto, "TCompactProtocol")) {
+        int size_type;
+        int state;
+        PyObject *globalVars;
+        PyObject *py_state;
+
+        globalVars = PyThreadState_GetDict();
+        py_state = PyDict_GetItemString(globalVars,"state");
+        if (!PyArg_Parse(py_state, "i", &state)) {
+            return ;
+        }
+
+        if (state != VALUE_READ && state != CONTAINER_READ) {
+            PyErr_SetString(PyExc_ValueError, "Wrong state");
+            return;
+        }
+        size_type = readByteC(data);
+        *size = size_type>>4;
+        *etype = getTType(size_type);
+        if(*size == 15) {
+            *size = readSize(data);
+        }
+        pushContainers(state);
+        PyDict_SetItemString(globalVars,"state",Py_BuildValue("i",CONTAINER_READ));
+    }
+}
+
+void
+readSetEnd(char *data, char *proto)
+{
+    if (!strcmp(proto, "TJSONProtocol")) {
+        readJSONArrayEnd(data);
+    }
+    if (!strcmp(proto, "TCompactProtocol")) {
+        int temp;
+        int state;
+        PyObject *globalVars;
+        PyObject *py_state;
+
+        globalVars = PyThreadState_GetDict();
+        py_state = PyDict_GetItemString(globalVars,"state");
+        if (!PyArg_Parse(py_state, "i", &state)) {
+            return ;
+        }
+        if (state != CONTAINER_READ) {
+            PyErr_SetString(PyExc_ValueError, "Wrong state");
+            return;
+        }
+        temp = popContainers();
+        PyDict_SetItemString(globalVars,"state",Py_BuildValue("i",0));
+    }
+    return;
+}
+
+void
+readListBegin(char *data, char *proto, int *etype, int *size)
+{
+    if (!strcmp(proto, "TBinaryProtocol")) {
+        *etype = readByteC(data);
+        *size = readI32(data, proto);
+    }
+    if (!strcmp(proto, "TJSONProtocol")) {
+        readJSONArrayStart(data);
+        *etype = JTypesToInt(readJSONString(data, 0));
+        *size = readJsonInteger(data);
+    }
+    if (!strcmp(proto, "TCompactProtocol")) {
+        int size_type;
+        int state;
+        PyObject *globalVars;
+        PyObject *py_state;
+
+        globalVars = PyThreadState_GetDict();
+        py_state = PyDict_GetItemString(globalVars,"state");
+        if (!PyArg_Parse(py_state, "i", &state)) {
+            return ;
+        }
+
+        if (state != VALUE_READ && state != CONTAINER_READ) {
+            PyErr_SetString(PyExc_ValueError, "Wrong state");
+            return;
+        }
+        size_type = readByteC(data);
+        *size = size_type>>4;
+        *etype = getTType(size_type);
+        if (*size == 15) {
+            *size = readSize(data);
+        }
+        pushContainers(state);
+        PyDict_SetItemString(globalVars,"state",Py_BuildValue("i",CONTAINER_READ));
+    }
+}
+
+void
+readListEnd(char *data, char *proto)
+{
+    if (!strcmp(proto, "TJSONProtocol")) {
+        readJSONArrayEnd(data);
+    }
+    if (!strcmp(proto, "TCompactProtocol")) {
```

Comment:
this should be else if to make it more clear that it's this branch or the previous one (not both)